### ISSUE TITLE
[ProxyManagerBridge] Invalid code generated for \Redis lazy service

### ISSUE THREAD

**Issue by scaytrase** [symfony/symfony#23066]
| Q                | A
| ---------------- | -----
| Bug report?      | yes
| Feature request? | no
| BC Break report? | no
| RFC?             | no
| Symfony version  | 3.2.7, 3.2.9

zendframework/zend-code         3.1.0 (2.6.3 fails too)
ocramius/proxy-manager    2.0.4 (1.0.2 fails too)

Given a code snippet

```php
namespace test;

use Symfony\Bridge\ProxyManager\LazyProxy\PhpDumper\ProxyDumper;
use Symfony\Component\DependencyInjection\Definition;

require_once __DIR__ . '/vendor/autoload.php';

$definition = new Definition(\Redis::class);

$dumper = new ProxyDumper();
var_dump($dumper->getProxyCode($definition));
```

It generates invalid methods like
```php
    /**
     * {@inheritDoc}
     */
    public function exists()
    {
        $this->initializer59355dd57400c356686291 && $this->initializer59355dd57400c356686291->__invoke($this->valueHolder59355dd573f67264382740, $this, 'exists', array(), $this->initializer59355dd57400c356686291);

        return $this->valueHolder59355dd573f67264382740->exists();
    }
```

They have no arguments and most of other methods are same (but not all)
Which make PHP immediately fail on `exist('test')` with

```
An exception has been thrown during the rendering of a template ("Warning: Redis::exists() expects exactly 1 parameter, 0 given").
```

If I generate the proxy and immediately use it like this, it works (dump false)

```php
$factory = new LazyLoadingValueHolderFactory($config);

/** @var \Redis|VirtualProxyInterface $proxy */
$proxy = $factory->createProxy(
    \Redis::class,
    function (& $wrappedObject, $proxy, $method, $parameters, & $initializer) {
        $wrappedObject = new \Redis(); // instantiation logic here
        $wrappedObject->connect('192.168.8.208');
        $initializer = null; // turning off further lazy initialization
    }
);

var_dump($proxy->exists('test'));
```

As soon as `ProxyManager` works well by itself, I suspect that problem is with the bridge, mut might be another library (i.e zendframework/code).

Actually this fails with the following service definition, so posting here first
```yaml
  bankiru.redis:
    class: Redis
    lazy: true
    calls:
    - [connect, ["%redis.host%", "%redis.port%"]]
```

**Comment by nicolas-grekas** [comment]
100% valid issue, here is the cause of it: https://github.com/phpredis/phpredis/issues/1092
Closing as we cant do anything about it.
Maybe add a special case for it in ProxyManager? I'll let you talk about it with @Ocramius on this repo :)

**Comment by scaytrase** [comment]
Hell, it's deeper that I think. Thank you for pointing, so I think it's easier for now to workaround than wait for fix :(

**Comment by Ocramius** [comment]
There are two possible solutions to this:

 1. fix `\Redis` reflection so that it reports correct reflection information (that's why I have a passionate hate for extensions btw: they never work properly like userland code)
 2. extend `\Redis` in userland, with all method definitions as in the documentation, then proxy that.

A special case for ProxyManager will not be considered.

Related: https://github.com/Ocramius/ProxyManager/issues/162
Related: https://github.com/Ocramius/ProxyManager/issues/369

**Comment by scaytrase** [comment]
@Ocramius is it possible to have another proxy factory, which generates calls with someting like 

```php
return call_user_func_array('parent::'.__METHOD__, func_get_args())
```

instead of direct call? And use it for predefined list of classes, i.e in a form of decorator around original factory?

**Comment by Ocramius** [comment]
> @Ocramius is it possible to have another proxy factory, which generates calls with someting like

```php
return call_user_func_array('parent::'.__METHOD__, func_get_args())
```

This has been considered before, but won't be implemented for a while, sorry. There are other implications in doing this (which I can't really remember ATM), so it was held off.

 > Userland extending for `\Redis` is a bit equivalent of implementing lazyness by-hand, as soon as almost none methods are properly reflected.

You don't need to implement laziness: just a no-op wrapper around a `Redis` instance

**Comment by Ocramius** [comment]
> > You don't need to implement laziness: just a no-op wrapper around a Redis instance
 > I didn't mean I have to, but there is a little step from it, so why not.

Lots to be taken in consideration there - it's not a walk in the park, especially testing-wise :-)

 > I understand that may have some performance issues, but since the code is generated and dumping (we are still discussing the symfony issue, I think), so having configurable in the bridge with empty list by default would have no impact if implemented as a part of the bridge

Could you rephrase? I didn't understand this part.

**Comment by scaytrase** [comment]
> Could you rephrase? I didn't understand this part.

Currently the bridge manages the symfony part of the proxy generation, especially these lines 

https://github.com/symfony/symfony/blob/master/src/Symfony/Bridge/ProxyManager/LazyProxy/PhpDumper/ProxyDumper.php#L130-L132 is for generating proxy

Making some parts of this code configurable we can create a `GeneratorInterface` which will generate mentioned above signature for parent call. 

Quick draft will look like

```php
    private static $nonReflectedClasses = [];
    /**
     * @return ClassGenerator
     */
    private function generateProxyClass(Definition $definition)
    {
        $generatedClass = new ClassGenerator($this->getProxyClassName($definition));
        if (!in_array($definition->getClass(), self::$nonReflectedClasses, true)) {
            // most of the time this will work
           $proxyGenerator = new LazyLoadingValueHolderGenerator();    
        } else {
            // but here we can try workaround non reflected classes
           $proxyGenerator = new CallUserFuncArrayHolderGenerator(); 
        }
        $proxyGenerator->generate(new \ReflectionClass($definition->getClass()), $generatedClass);
        return $generatedClass;
    }

   public static function addNonReflectedClass($className)
   {
       self::$nonReflectedClasses[] = $className;
   }
```

https://github.com/Ocramius/ProxyManager/blob/1.0.x/src/ProxyManager/ProxyGenerator/LazyLoadingValueHolder/MethodGenerator/LazyLoadingMethodInterceptor.php#L60-L67

I think we wil need just use another interceptor (not familiar with code, might be wrong place)

Static is just dirty example, as well as creating generator for each dumped class.

Since this works only during container compilation we will overcome the problem with no performance impact for users who does not suffer this issue, but at least we can allow create proxy for every class, however with possibly degraded performance (I suppose that `call_user_func_array` should be a bit slower than super call)

This solution will allow us do the following

```

$definition = new Definition(\Redis::class);

$dumper = new ProxyDumper();
ProxyDumper::addNonReflectedClass(\Redis::class);
var_dump($dumper->getProxyCode($definition));
```

And things should go fine

**Comment by Ocramius** [comment]
> Making some parts of this code configurable we can create a `GeneratorInterface` which will generate mentioned above signature for parent call.

and

 > I think we wil need just use another interceptor (not familiar with code, might be wrong place)

Yes, that is what I mean by "This has been considered before". It's not going to happen in the near future, as I won't have time to work on it for now. If that method (`call_user_func()` + `func_get_args()`) ends up being more reliable, it will be the default approach. This has nothing to do with "configurability" of symfony, but rather with just the defaults of `ProxyManager`.

**Comment by scaytrase** [comment]
Okay, thank you for explanation
